Rya 451 Query manager

isper3at opened the pull request:
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->
## Description
>What Changed?
[Brief Description of what changed]

### Tests
>Coverage?

a list of tests to be written

### Links
[Jira](https://issues.apache.org/jira/browse/RYA-451)

### Checklist
- [ ] Code Review
- [x] Squash Commits

#### People To Reivew
@kchilton2

isper3at committed:
checkpoint

isper3at committed:
checkpoint

isper3at committed:
RYA-451 QueryManage

No tests exist yet
Logs and console output needs to be cleaned up

kchilton2 commented on extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/entity/StreamsQuery.java:
```diff
@@ -39,9 +37,12 @@
-     * @param queryId - Uniquely identifies the query within Rya Streams. (not null)
-     * @param sparql - The SPARQL query that defines how statements will be processed. (not null)
-     * @param isActive - {@code true} if Rya Streams should process this query; otherwise {@code false}.
+     * @param queryId - Uniquely identifies the query within Rya Streams. (not
```
Rogue formatter is changing all the spacing. Please don't do that. It makes reviewing your code take a lot longer than it needs to.

isper3at replied:
yeah I saw that, I put this up on the way out the door so I didn't stop to see this kinda stuff

kchilton2 commented on extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/QueryManagerIT.java:
```diff
@@ -0,0 +1,38 @@
+
+import org.junit.Before;
+
+public class QueryManagerIT {
```
You shouldn't need to write integration test to test the query manager. You can mock everything and just write unit tests.

isper3at replied:
Yeah, I named it before i thought about what was going to need to be tested and how

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+public class QueryManager extends AbstractIdleService {
+    private static final Logger LOG = LoggerFactory.getLogger(QueryManager.class);
+
+    private final String ryaInstanceName;
```
This manager needs to manage all of the rya instances maintained by the Rya Streams system, so I don't know why this field exists.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+     * @param scheduler - The {@link Scheduler} to use throughout
+     */
+    public QueryManager(final String ryaInstanceName, final QueryExecutor queryExecutor,
+            final Map<String, QueryChangeLogSource> sources, final Set<QueryRepository> queryRepos,
```
You shouldn't be passing query repositories into this constructor since all of them have to be discovered through the QueryChangeLogSources

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+        this.queryRepos = requireNonNull(queryRepos);
+        this.scheduler = requireNonNull(scheduler);
+
+        // subscribe to the repos and sources to be notified of changes.
```
This happens only when a source discovers a change log.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+    private void runQuery(final StreamsQuery query) {
+        requireNonNull(query);
+        LOG.trace("Starting Query: " + query.getSparql());
+        if (!queryExecutor.isRunning()) {
```
This case will never happen since you start it in this service's startup method.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+    @Override
+    protected void startUp() throws Exception {
+        LOG.trace("Starting Query Manager.");
+        queryRepos.forEach(repo -> repo.startAndWait());
```
This object is going away, so you don't need this line.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+     * @param scheduler - The {@link Scheduler} to use throughout
+     */
+    public QueryManager(final String ryaInstanceName, final QueryExecutor queryExecutor,
+            final Map<String, QueryChangeLogSource> sources, final Set<QueryRepository> queryRepos,
```
Also not sure why the sources is a map. What's the key? I'm pretty sure this should only be a Set of sources.

isper3at replied:
Rya Instance name.  I needed a way of determining and hanging onto the rya instance name of each source

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+    protected void startUp() throws Exception {
+        LOG.trace("Starting Query Manager.");
+        queryRepos.forEach(repo -> repo.startAndWait());
+        sources.forEach((ryaInstance, source) -> source.startAndWait());
```
Each source provides many rya instances, not just one.

isper3at replied:
so multimap source -> rya instance names?  I am ultimately ignoring rya instance here though

isper3at replied:
the big problem I'm having is I need a way of getting, maintaining the rya instance names

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+     */
+    private class QueryManagerSourceListener implements SourceListener {
+        @Override
+        public void notifyCreate(final String ryaInstanceName, final QueryChangeLog log) {
```
This method needs an info log saying "Discovered new Query Change Log for Rya Instance [instance name here] within source [source identifying information here]."

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+        }
+
+        @Override
+        public void notifyDelete(final String ryaInstanceName) {
```
This method needs an info log similar to the found log I mentioned.

kchilton2 replied:
This method also needs to shutdown the repository that was made for the rya instance and remove it from the repos list.

isper3at replied:
that's what I need to hang onto rya instance name references

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+ * instances/rya streams instances.
+ */
+@DefaultAnnotation(NonNull.class)
+public class QueryManager extends AbstractIdleService {
```
This whole class has thread safety concerns that need to be addressed.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    @Override
+    protected void shutDown() throws Exception {
```
Might want to use stop and wait calls instead of just stop to prevent this method from returning until everything the class depends on finishes stopping.

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+                                entry.getIsActive().or(false));
+                        runQuery(newQuery);
+                        queriesCache.put(entry.getQueryId(), newQuery);
+                        LOG.trace("Created new query: " + newQuery.toString());
```
I'm not sure how you're choosing to log at debug and trace, but it seems like all of this should be debug. We also should print an info whenever we're changing query execution stuff in general, otherwise it's going to be hard to tell that this process is doing anything when deployed.

isper3at replied:
yeah got trace/info mixed up.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+     * to stop the query, stops the query. Otherwise, if the query is not
+     * running, it is removed.</li>
+     */
+    private class QueryManagerQueryChange implements QueryChangeLogListener {
```
Seeing how we implemented this class kind of makes me think this interface sucks. We could simplify this code a lot if the listener also provided the updated copy of the StreamsQuery since that's all you really need to do your work in here. I think the listener is registered with the repository, we the repsitory is already applying this logic.

isper3at replied:
that would make things easier, but then the repos would have to know about a query executor...we can talk more tomorrow

kchilton2 commented on extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/QueryManagerIT.java:
```diff
@@ -0,0 +1,38 @@
+        //
+    }
+
+    // Test No query change log sources, add a new one
```
I don't think there is a way to add a new one with the current implementation.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+    private final Set<QueryRepository> queryRepos;
+
+    private final Map<String, QueryChangeLogSource> sources;
+    private final Map<UUID, StreamsQuery> queriesCache;
```
This shouldn't be needed since the repository knows this information already. This kind of indicates a problem with the listener API.

kchilton2 commented on extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/QueryManagerIT.java:
```diff
@@ -0,0 +1,38 @@
+
+    // Test No query change log sources, add a new one
+
+    // Test existing query change log sources, add a new one with same rya instance
```
I don't think this test makes sense either since you can't know which rya instances are in each source. This does bring up an interesting problem though. I hadn't thought about seeing the same rya instance in more than a single source.

kchilton2 commented on extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/QueryManagerIT.java:
```diff
@@ -0,0 +1,38 @@
+
+    // Test existing query change log sources, add a new one with same rya instance
+    
+    // Test No query change log sources, remove one
```
This case is only in the constructor.

isper3at replied:
I should change this to say QueryChangeLogSource with no QueryChangeLogs and a notification with a remove gets triggered

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,241 @@
+            LOG.debug("Notified of new QueryChangeLog, creating new repository");
+            final QueryRepository repo = new InMemoryQueryRepository(log, scheduler);
+            repo.subscribe(new QueryManagerQueryChange());
+            repo.startAndWait();
```
This hunk of code will not work. We need to start the query repository prior to subscribing to it, otherwise the listener will iterate through the entire list of changes and that may cause it to create queries to run that are later deleted, start and stop things multiple times, etc. That would be bad. Subscribe returns the current state of the queries so that the current state can be checked against the executing state (which is why the executor has a method for checking quick queries are running).

isper3at committed:
checkpoint

isper3at committed:
Review checkpoint

Altered listener to return the updated query
updated query manager to not cache query
cache repository with rya instance name

added reentrant lock, but haven't used it yet.

isper3at committed:
Added reentrant lock use

isper3at committed:
started tests

isper3at committed:
Added tests

kchilton2 commented on extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/queries/InMemoryQueryRepository.java:
```diff
@@ -228,13 +228,16 @@ private void updateCache() {
                 final QueryChange change = entry.getEntry();
                 final UUID queryId = change.getQueryId();
 
+                StreamsQuery newQueryState = null;
```
All of the changes in this file could be replaced with these two lines at the end:
```
Optional<StreamsQuery> newQueryState = Optional.ofNullable( queriesCache.get(queryId) );
listeners.forEach(listener -> listener.notify(entry, newQueryState ));
```

isper3at replied:
ok

kchilton2 commented on extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/queries/QueryChangeLogListener.java:
```diff
@@ -36,6 +40,7 @@
      * on the repository.
      *
      * @param queryChangeEvent - The event that occurred. (not null)
+     * @param newQueryState - The new state of the query after the query change event. (not null)
```
Indicate when this is not present. I assume it's whenever the change is a DELETE.

isper3at replied:
done

kchilton2 commented on extras/rya.streams/api/src/test/java/org/apache/rya/streams/api/queries/InMemoryQueryRepositoryTest.java:
```diff
@@ -167,7 +167,7 @@ public void updateListenerNotify() throws Exception {
             final Set<StreamsQuery> existing = queries.subscribe(new QueryChangeLogListener() {
                 @Override
-                public void notify(final ChangeLogEntry<QueryChange> queryChangeEvent) {
+                public void notify(final ChangeLogEntry<QueryChange> queryChangeEvent, final Optional<StreamsQuery> newQueryState) {
```
The tests in this class need to be updated to ensure they are notified with the correct newQueryState value.

kchilton2 replied:
This comment has not been addressed.

kchilton2 commented on extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/QueryManagerTest.java:
```diff
@@ -0,0 +1,185 @@
+
+import com.google.common.util.concurrent.AbstractScheduledService.Scheduler;
+
+public class QueryManagerTest {
```
docs

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/QueryManagerTest.java:
```diff
@@ -0,0 +1,185 @@
+package org.apache.rya.streams.querymanager;
```
license

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+    /**
+     * Creates a new {@link QueryManager}.
+     *
+     * @param queryExecutor - The {@link QueryExecutor} that starts/creates
```
This is kind of an awkward doc. I think the details of when stuff is started/stopped should be at the class level. Here you could just say, ```"Runs the active {@link StreamsQuery}s. (not null)```

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+     *        queries when a CREATED query is found in the
+     *        {@link QueryChangeLog}. (not null)
+     * @param source - The {@link QueryChangeLogSource} of QueryChangeLogs. (not null)
+     * @param scheduler - The {@link Scheduler} to use throughout
```
Indicate what this is used for. It's kind of weird to pass a scheduler into an idle service, so mention it's used to discover query changes within the logs or something.

isper3at replied:
I clearly started  doc-ing and just forgot to finish

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+        this.queryExecutor = requireNonNull(queryExecutor);
+        this.scheduler = requireNonNull(scheduler);
+
+        queryRepos = new HashMap<>();
```
This and lock could just be assigned outside of the constructor.

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+     * @param ryaInstanceName - The Rya instance the query belongs to.
+     * @param query - The query to run.(not null)
+     */
+    private void runQuery(final String ryaInstanceName, final StreamsQuery query) {
```
missing null check

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+     */
+    private void runQuery(final String ryaInstanceName, final StreamsQuery query) {
+        requireNonNull(query);
+        LOG.info("Starting Query: " + query.getSparql());
```
I'd like it if we included the rya instance name, query id, and the sparql when logging that a query is being started.

isper3at replied:
that's why I added a toString() to StreamsQuery

kchilton2 replied:
But you're just printing the SPARQL here.

kchilton2 replied:
And Rya Instance name isn't in StreamsQuery.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+    private void runQuery(final String ryaInstanceName, final StreamsQuery query) {
+        requireNonNull(query);
+        LOG.info("Starting Query: " + query.getSparql());
+        if (!queryExecutor.isRunning()) {
```
I don't think you need this check since the query executor throws this exception already if you try to use it while it isn't running.

isper3at replied:
removed

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+        try {
+            queryExecutor.startQuery(ryaInstanceName, query);
+        } catch (final QueryExecutorException e) {
+            LOG.error("Failed to start query.", e);
```
So we just give up and never try to start it again? I'm not sure what we should be doing here.

isper3at replied:
I mean we could have a thread that occasionally tries to start it until it succeeds, but that also doesn't seem right.  The biggest problem is that if this does fail and logs, if no one is watching the logs there's no way to know it failed.  A claxon?

kchilton2 replied:
I think we're kind of covered by having the QueryExecutor be responsible for managing what is started and stopped. So really, it throwing the exception indicates it wasn't able to even process the start request. But we haven't really formalized that yet beyond being able to ask the executor what is started. Maybe we should think a little more on what the executor's responsibilities are.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+        requireNonNull(queryId);
+
+        LOG.info("Stopping query: " + queryId.toString());
+        if (!queryExecutor.isRunning()) {
```
This method gets the same comments as the start method.

isper3at replied:
removed

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+        public void notifyCreate(final String ryaInstanceName, final QueryChangeLog log) {
+            lock.lock();
+            try {
+                LOG.info("Discovered new Query Change Log for Rya Instance " + ryaInstanceName + " within source " + log.toString());
```
QueryChangeLog's toString() is overridden to be anything meaningful. This is just going to be a hash. That isn't useful. We only have one source right now, so I think you can just assume which source it came from. We could log that configuration information at startup of the daemon.

isper3at replied:
So I won't log the source

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+                LOG.info("Discovered new Query Change Log for Rya Instance " + ryaInstanceName + " within source " + log.toString());
+                final QueryRepository repo = new InMemoryQueryRepository(log, scheduler);
+                repo.startAndWait();
+                repo.subscribe(new QueryManagerQueryChange(ryaInstanceName));
```
Doesn't this return a list StreamsQuery objects that need to be handled?

isper3at replied:
yes....

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+                final QueryRepository repo = new InMemoryQueryRepository(log, scheduler);
+                repo.startAndWait();
+                repo.subscribe(new QueryManagerQueryChange(ryaInstanceName));
+                LOG.info("New query repository started");
```
I don't think we need this log to be info.

isper3at replied:
removed

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+        public void notifyDelete(final String ryaInstanceName) {
+            lock.lock();
+            try {
+                LOG.info("Notified of deleting QueryChangeLog, stopping all queries belonging to the change log.");
```
Indicate which Rya Instance it was for.

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+    private class QueryManagerQueryChange implements QueryChangeLogListener {
+        private final String ryaInstanceName;
+
+        public QueryManagerQueryChange(final String ryaInstanceName) {
```
docs

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+                switch (entry.getChangeType()) {
+                    case CREATE:
+                        LOG.debug("Creating query event.");
+                        if (newQueryState.isPresent()) {
```
What if it isn't present?

isper3at replied:
It's past.

I'll log an error?

kchilton2 replied:
What do you mean by it's past?

kchilton2 replied:
Oh, you mean ignored? Yea, we should probably at least log something.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+                        LOG.debug("Creating query event.");
+                        if (newQueryState.isPresent()) {
+                            runQuery(ryaInstanceName, newQueryState.get());
+                            LOG.info("Created new query: " + newQueryState.get().toString());
```
I'm not sure we need logs in this listener since all the methods they call do logging anyway.

isper3at replied:
fair enough

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+                                        entry.getIsActive().or(false));
+                                runQuery(ryaInstanceName, newQuery);
+                                LOG.info("Starting query: " + newQuery.toString());
+                            } else if (!updateQuery.isActive() && !entry.getIsActive().or(true)) {
```
I don't think this logic is correct. You don't have enough information within this conditional to know if a change happened since you don't have the state prior to the change being applied. You know what the change was and the new state of the query after the change. You could ask the executor if it is currently running that query and stop it, though.

isper3at replied:
so, if the new state is active, just try having the executor start/stop it?

kchilton2 replied:
I don't honestly see anything wrong with that now that I think about it. The executor could be responsible for ignoring repetitive requests since it knows what is running anyway.

kchilton2 replied:
Which kind of makes me think the edit we made for supplying the finished state of the query might not actually be needed.

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,263 @@
+                            final StreamsQuery updateQuery = newQueryState.get();
+                            // if the query is currently inactive, and is updated to
+                            // be active, turn on.
+                            if (!updateQuery.isActive() && entry.getIsActive().or(false)) {
```
Kind of the same comment here. You don't know the state before the change was applied within this conditional, so you can't know if you need to start it based on this.

isper3at committed:
responded to review

isper3at committed:
Added conditional check and error logging

kchilton2 commented on extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/QueryManagerTest.java:
```diff
@@ -0,0 +1,204 @@
+
+        // when the query executor is told to start the test query on the test
+        // rya instance, count down on the countdown latch
+        Mockito.doAnswer(invocation -> {
```
Could static import all of the Mockito functions.

isper3at replied:
addressed

isper3at committed:
updated test

kchilton2 commented on extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/QueryManagerTest.java:
```diff
@@ -0,0 +1,204 @@
+    public static void setupLogger() {
+        org.apache.log4j.Logger.getRootLogger().addAppender(new ConsoleAppender());
+        LOG = LoggerFactory.getLogger(QueryManagerTest.class);
+        LOG.info("test");
```
This is unnecessary.

isper3at replied:
addressed

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,262 @@
+    private void runQuery(final String ryaInstanceName, final StreamsQuery query) {
+        requireNonNull(ryaInstanceName);
+        requireNonNull(query);
+        LOG.info("Starting Query: " + query.toString());
```
Include the rya instance name in this log.

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,262 @@
+     * to stop the query, stops the query. Otherwise, if the query is not
+     * running, it is removed.</li>
+     */
+    private class QueryManagerQueryChange implements QueryChangeLogListener {
```
This class has kind of a weird name.

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,262 @@
+                        break;
+                    case UPDATE:
+                        if (!newQueryState.isPresent()) {
+                            LOG.debug("The query with ID: " + entry.getQueryId() + " must be provided with the update, cannot perform update.");
```
This log doesn't reflect your is present check.

isper3at replied:
done

kchilton2 commented on extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java:
```diff
@@ -0,0 +1,262 @@
+                            final StreamsQuery updateQuery = newQueryState.get();
+                            // if the query is currently inactive, and is updated to
+                            // be active, turn on.
+                            if (!updateQuery.isActive() && entry.getIsActive().or(false)) {
```
Thought you were going to change the start/stop logic for updates.

isper3at replied:
addressed

isper3at committed:
responded to comment

isper3at committed:
updated
